doc/design/design.md

```diff
@@ -0,0 +1,228 @@
+# Design
+
+<!-- 
+INFO: This file contains nested plantuml diagrams. For displaying them you need to enable editor support.
+For InteliJ see: https://youtrack.jetbrains.com/api/files/74-1236265?sign=MTY1MDg0NDgwMDAwMHwxMS0xfDc0LTEyMzYyNjV8d2VXdWNFVkJfVWZQeV9ZX3JHNF9rYno4cFV2%0D%0AMy1BV2ZSNjZaOUxhdVBOSQ0K%0D%0A&updated=1639999300525
+For VS Code: https://marketplace.visualstudio.com/items?itemName=jebbs.plantuml
+-->
+
+## Overall Architecture
+
+The extension installer is split into a backend (running in the database) and a client (running in the Saas ui).
+
+```plantuml
+@startuml
+node ExasolCluster{
+rectangle backend as "extension manager backend(go)"{
+rectangle JsVM
+}
+storage Extensions{
+file installer as "postgres-extension-def-1.0.0.js"
+}
+}
+
+node Browser{
+rectangle jsClient{
+
+    }
+}
+
+jsClient <---> backend
+JsVM <- installer: "loaded into"
+@enduml
+```
+
+### Do we need a backend?
+
+One option would be to implement everything in the JS client. However, we discarded that option, since it does not allow
+us upgrade the installed adapters automatically.
+
+### Extensions
+
+The extension manager has an extension mechanism.
+The extension manager itself has no dependencies to the virtual schema projects.
+They are loaded at runtime.
+
+```plantuml	
+postgres as [PostgreSQL Virtual Schema\nExtension definition]
+mysql as [MySQL Virtual Schema\nExtension definition]
+extensionInterface as [Extension manger interface]
+
+extensionInterface <-- [Extension manager]
+extensionInterface <|-- postgres 
+extensionInterface <|-- mysql
+
+[Extension manager] -> postgres: loaded at runtime
+[Extension manager] -> mysql: loaded at runtime
+```
+
+## Extension API
+
+We have one extension definition per extension (integration project).
+All extension definitions implement the same public interface:
+
+```plantuml
+interface ExasolExtension{
+
+}
+
+ExasolExtension <-- "postgres-virtual-schema (repo).postgres-extension-definition"
+ExasolExtension <-- "mysql-virtual-schema (repo).MySqlExtensionDefinition"
+```
+
+### Language
+
+For a language we decided for TypeScript.
+
+Alternative JSON: We decided for a programming language like JS since some extensions have more complex installation
+mechanisms. Expressing this by just providing SQL scripts with placeholders quickly reaches its limit.
+Especially the function for reading back parameter values can get a bit more complex.
+
+Alternative Lua: We could also use Lua. We decided for JS since:
+
+* JS it more popular --> more developers know it
+* JS is used in the frontend. If at some point we change our decision and want to run the installation in the client, we
+  don't need to update the installers.
+
+The compiled JavaScript code is evaluated in the backend using a nested JavaScript interpreter.
+By that it's isolated and for example can't access resources like disc or network.
```

Comment:
What are the capabilities of the isolated interpreter? Is it simply running JS code without any I/O?

Reply:
Who can create extensions?

Reply:
How do we get the files in place, or did we "outsource" this?

Reply:
> hat are the capabilities of the isolated interpreter? Is it simply running JS code without any I/

yes.

Reply:
> do we get the files in place, or did we "outsource" this?

for now it's outsourced

Reply:
> an create extensions?

answered